File: destack-py/destack/encoder/json/wiring.py
import base64
from datetime import UTC, date, datetime, time
from typing import TYPE_CHECKING, Any, assert_never

from destack.language.core import (
    BuiltinObject,
    Encoding,
    Json,
    NodeReference,
    NodeType,
    PrimitiveType,
    ScalarType,
    Type,
    TypeCardinality,
)
from destack.language.registry import (
    ENUM_CLASS_BY_TYPE,
    NODE_CLASS_BY_TYPE,
    STRUCT_CLASS_BY_TYPE,
)
from destack.utils.log import get_logger
from destack.utils.telemetry import get_tracer
from destack.utils.time import timedelta_from_isoformat, timedelta_to_isoformat
from destack.utils.uuid import UUID

if TYPE_CHECKING:
    from destack.language import Session


# pyright: reportIncompatibleVariableOverride=false


logger = get_logger(__name__)
tracer = get_tracer(__name__)
type_ = type


def pack_json(value: Any, type: Type) -> Json:
    """Pack a generic typed value to a JSON object."""
    if type.cardinality == TypeCardinality.SCALAR:
        if value is None:
            return None
        else:
            return _pack_scalar_json(value, type)
    elif type.cardinality == TypeCardinality.LIST:
        if value is None:
            return None
        else:
            packed_list: list[Any] = []
            for item in value:
                packed_list.append(_pack_scalar_json(item, type))
            return packed_list
    elif type.cardinality == TypeCardinality.MAP:
        if value is None:
            return None
        else:
            assert type.key_type is not None, f"no key type for {type!r}"
            packed_map: dict[str, Any] = {}
            for key, val in value.items():
                # always use string keys in JSON
                packed_key = str(_pack_scalar_json(key, type.key_type))
                packed_val = _pack_scalar_json(val, type)
                packed_map[packed_key] = packed_val
            return packed_map
    else:
        assert_never(type.cardinality)


def unpack_json(value: Json, type: Type, session: "Session | None") -> Any:
    """Unpack a JSON object to a generic typed value."""
    if type.cardinality == TypeCardinality.SCALAR:
        if value is None:
            return None
        else:
            return _unpack_scalar_json(value, type, session)
    elif type.cardinality == TypeCardinality.LIST:
        if value is None:
            return None
        else:
            unpacked_list = []
            for item in value:
                unpacked_list.append(_unpack_scalar_json(item, type, session))
            return unpacked_list
    elif type.cardinality == TypeCardinality.MAP:
        if value is None:
            return None
        else:
            unpacked_map = {}
            assert isinstance(value, dict), f"expected dict for map type, got {type_(value)}"
            for key, val in value.items():
                unpacked_key = (
                    _unpack_scalar_json(key, type.key_type, session) if type.key_type else key
                )
                unpacked_val = _unpack_scalar_json(val, type, session)
                unpacked_map[unpacked_key] = unpacked_val
            return unpacked_map
    else:
        assert_never(type.cardinality)


def _pack_scalar_json(value: Any, type: Type) -> Any:
    """Pack a scalar value to JSON."""
    if type.scalar_type == ScalarType.PRIMITIVE:
        assert type.primitive_type is not None, f"no primitive type for {type!r}"
        if type.primitive_type == PrimitiveType.BOOLEAN:
            return value
        elif (
            type.primitive_type
            in (
                PrimitiveType.SINT8,
                PrimitiveType.SINT16,
                PrimitiveType.SINT32,
                PrimitiveType.SINT64,
                PrimitiveType.SINT128,
            )
            or type.primitive_type
            in (
                PrimitiveType.UINT8,
                PrimitiveType.UINT16,
                PrimitiveType.UINT32,
                PrimitiveType.UINT64,
                PrimitiveType.UINT128,
            )
            or type.primitive_type
            in (
                PrimitiveType.FLOAT16,
                PrimitiveType.FLOAT32,
                PrimitiveType.FLOAT64,
            )
        ):
            return float(value)
        elif type.primitive_type == PrimitiveType.DATETIME:
            return value.astimezone(UTC).isoformat()
        elif type.primitive_type == PrimitiveType.DATE:
            return value.isoformat()
        elif type.primitive_type == PrimitiveType.TIME:
            return value.astimezone(UTC).replace(tzinfo=None).isoformat()
        elif type.primitive_type == PrimitiveType.DURATION:
            return timedelta_to_isoformat(value)
        elif type.primitive_type == PrimitiveType.STRING:
            return value
        elif type.primitive_type == PrimitiveType.UUID:
            return str(value)
        elif type.primitive_type == PrimitiveType.BYTES:
            return base64.b64encode(value).decode()
        elif type.primitive_type == PrimitiveType.JSON:
            return value
        else:
            assert_never(type.primitive_type)
    elif type.scalar_type == ScalarType.ENUM:
        # use the enum name in CONSTANT_UPPER_CASE for JSON
        return value.name
    elif type.scalar_type in (ScalarType.NODE_REFERENCE, ScalarType.NODE_VALUE, ScalarType.STRUCT):
        assert isinstance(value, BuiltinObject), (
            f"expected BuiltinObject for {type!r}, got {value!r}"
        )
        return value.pack(Encoding.JSON)
    else:
        assert_never(type.scalar_type)


def _unpack_scalar_json(value: Any, type: Type, session: "Session | None") -> Any:
    """Unpack a scalar value from JSON."""
    if type.scalar_type == ScalarType.PRIMITIVE:
        assert type.primitive_type is not None, f"no primitive type for {type!r}"
        if type.primitive_type == PrimitiveType.BOOLEAN:
            return value
        elif type.primitive_type in (
            PrimitiveType.SINT8,
            PrimitiveType.SINT16,
            PrimitiveType.SINT32,
            PrimitiveType.SINT64,
            PrimitiveType.SINT128,
        ) or type.primitive_type in (
            PrimitiveType.UINT8,
            PrimitiveType.UINT16,
            PrimitiveType.UINT32,
            PrimitiveType.UINT64,
            PrimitiveType.UINT128,
        ):
            return int(value)
        elif type.primitive_type in (
            PrimitiveType.FLOAT16,
            PrimitiveType.FLOAT32,
            PrimitiveType.FLOAT64,
        ):
            return float(value)
        elif type.primitive_type == PrimitiveType.DATETIME:
            return datetime.fromisoformat(value).astimezone(UTC)
        elif type.primitive_type == PrimitiveType.DATE:
            return date.fromisoformat(value)
        elif type.primitive_type == PrimitiveType.TIME:
            return time.fromisoformat(value).replace(tzinfo=None)
        elif type.primitive_type == PrimitiveType.DURATION:
            return timedelta_from_isoformat(value)
        elif type.primitive_type == PrimitiveType.STRING:
            return value
        elif type.primitive_type == PrimitiveType.UUID:
            return UUID(value)
        elif type.primitive_type == PrimitiveType.BYTES:
            return base64.b64decode(value)
        elif type.primitive_type == PrimitiveType.JSON:
            return value
        else:
            assert_never(type.primitive_type)
    elif type.scalar_type == ScalarType.ENUM:
        assert type.enum_type is not None, f"no enum type for {type!r}"
        enum_cls = ENUM_CLASS_BY_TYPE[type.enum_type]
        # enum name is in CONSTANT_UPPER_CASE
        return enum_cls[value]
    elif type.scalar_type == ScalarType.NODE_REFERENCE:
        return NodeReference.unpack(Encoding.JSON, value, session)
    elif type.scalar_type == ScalarType.NODE_VALUE:
        # metatype is stored as the enum name
        node_type = NodeType[value["metatype"]]
        node_cls = NODE_CLASS_BY_TYPE[node_type]
        return node_cls.unpack(Encoding.JSON, value, session)
    elif type.scalar_type == ScalarType.STRUCT:
        assert type.struct_type is not None, f"no struct type for {type!r}"
        struct_cls = STRUCT_CLASS_BY_TYPE[type.struct_type]
        return struct_cls.unpack(Encoding.JSON, value, session)
    else:
        assert_never(type.scalar_type)

File: destack-ts/src/encoder/json/wiring.ts
import { BuiltinObject } from "@destack/language/core/builtin";
import {
  Encoding,
  type NodeType,
  PrimitiveType,
  ScalarType,
  StructType,
  TypeCardinality,
} from "@destack/language/core/builtin/common";
import type { PropertyDefinition } from "@destack/language/core/builtin/definition";
import type { NodeReference } from "@destack/language/core/builtin/relation";
import type { Type } from "@destack/language/core/builtin/type";
import type { CustomProperty } from "@destack/language/core/common/property";
import type { Session } from "@destack/language/core/runtime/session";
import { NODE_CLASS_BY_TYPE, STRUCT_CLASS_BY_TYPE } from "@destack/language/registry";
import {
  assertNever,
  base64Decode,
  base64Encode,
  timedeltaFromISOFormat,
  timedeltaToISOFormat,
} from "@destack/utils";
import { Temporal } from "temporal-polyfill";

/**
 * Pack a generic typed value to a JSON object.
 */
export function packJson(value: any, type: Type | PropertyDefinition | CustomProperty): any {
  if (type.cardinality == TypeCardinality.SCALAR) {
    return _packScalarJson(value, type);
  } else if (type.cardinality == TypeCardinality.LIST) {
    if (!value) {
      return [];
    }
    const packedList: any[] = [];
    for (const item of value) {
      packedList.push(_packScalarJson(item, type));
    }
    return packedList;
  } else if (type.cardinality == TypeCardinality.MAP) {
    if (!value) {
      return {};
    }
    if (type.keyType === null) {
      throw new Error(`no key type for ${type.repr()}`);
    }
    const packedMap: { [key: string]: any } = {};
    for (const [key, val] of Object.entries(value)) {
      const packedKey = _packScalarJson(key, type.keyType);
      const packedVal = _packScalarJson(val, type);
      packedMap[String(packedKey)] = packedVal;
    }
    return packedMap;
  } else {
    assertNever(type.cardinality);
  }
}

/**
 * Unpack a JSON object to a generic typed value.
 */
export function unpackJson(
  value: any,
  type: Type | PropertyDefinition | CustomProperty,
  _session: Session | null,
): any {
  if (type.cardinality == TypeCardinality.SCALAR) {
    return _unpackScalarJson(value, type, _session);
  } else if (type.cardinality == TypeCardinality.LIST) {
    if (value === null) {
      return [];
    }
    const unpackedList = [];
    for (const item of value) {
      unpackedList.push(_unpackScalarJson(item, type, _session));
    }
    return unpackedList;
  } else if (type.cardinality == TypeCardinality.MAP) {
    if (value === null) {
      return {};
    }
    const unpackedMap: { [key: string]: any } = {};
    for (const [key, val] of Object.entries(value)) {
      const unpackedKey = type.keyType ? _unpackScalarJson(key, type.keyType, _session) : key;
      const unpackedVal = _unpackScalarJson(val, type, _session);
      unpackedMap[unpackedKey] = unpackedVal;
    }
    return unpackedMap;
  } else {
    assertNever(type.cardinality);
  }
}

/** Pack a scalar value to a JSON object. */
function _packScalarJson(value: any, type: Type | PropertyDefinition | CustomProperty): any {
  if (type.scalarType == ScalarType.PRIMITIVE) {
    if (type.primitiveType === null) {
      throw new Error(`missing primitive type for ${type.repr()}`);
    } else if (type.primitiveType == PrimitiveType.BOOLEAN) {
      return value;
    } else if (type.primitiveType == PrimitiveType.STRING) {
      return value;
    } else if (type.primitiveType == PrimitiveType.UUID) {
      return value;
    } else if (type.primitiveType == PrimitiveType.BYTES) {
      return base64Encode(value as Uint8Array);
    } else if (type.primitiveType == PrimitiveType.DATETIME) {
      return (value as Temporal.ZonedDateTime).toString({ timeZoneName: "never" });
    } else if (type.primitiveType == PrimitiveType.DATE) {
      return (value as Temporal.PlainDate).toString();
    } else if (type.primitiveType == PrimitiveType.TIME) {
      return (value as Temporal.PlainTime).toString();
    } else if (type.primitiveType == PrimitiveType.DURATION) {
      return timedeltaToISOFormat(value as Temporal.Duration);
    } else if (
      type.primitiveType == PrimitiveType.SINT8 ||
      type.primitiveType == PrimitiveType.SINT16 ||
      type.primitiveType == PrimitiveType.SINT32 ||
      type.primitiveType == PrimitiveType.SINT64 ||
      type.primitiveType == PrimitiveType.SINT128 ||
      type.primitiveType == PrimitiveType.UINT8 ||
      type.primitiveType == PrimitiveType.UINT16 ||
      type.primitiveType == PrimitiveType.UINT32 ||
      type.primitiveType == PrimitiveType.UINT64 ||
      type.primitiveType == PrimitiveType.UINT128 ||
      type.primitiveType == PrimitiveType.FLOAT16 ||
      type.primitiveType == PrimitiveType.FLOAT32 ||
      type.primitiveType == PrimitiveType.FLOAT64
    ) {
      return value;
    } else if (type.primitiveType == PrimitiveType.JSON) {
      return value;
    } else {
      assertNever(type.primitiveType);
    }
  } else if (type.scalarType == ScalarType.ENUM) {
    return value;
  } else if (
    type.scalarType == ScalarType.NODE_REFERENCE ||
    type.scalarType == ScalarType.NODE_VALUE ||
    type.scalarType == ScalarType.STRUCT
  ) {
    if (!(value instanceof BuiltinObject)) {
      throw new Error(
        `expected BuiltinObject for ${type.repr()}, got ${value.constructor.name}: ${value}`,
      );
    }
    return (value as BuiltinObject).pack(Encoding.JSON);
  } else {
    assertNever(type.scalarType);
  }
}

/** Unpack a JSON object to a scalar value. */
function _unpackScalarJson(
  value: any,
  type: Type | PropertyDefinition | CustomProperty,
  _session: Session | null,
): any {
  if (type.scalarType == ScalarType.PRIMITIVE) {
    if (type.primitiveType === null) {
      throw new Error(`missing primitive type for ${type.repr()}`);
    } else if (type.primitiveType == PrimitiveType.BOOLEAN) {
      return value;
    } else if (type.primitiveType == PrimitiveType.BYTES) {
      return base64Decode(value);
    } else if (type.primitiveType == PrimitiveType.UUID) {
      return value;
    } else if (type.primitiveType == PrimitiveType.STRING) {
      return value;
    } else if (type.primitiveType == PrimitiveType.DATETIME) {
      return Temporal.Instant.from(value).toZonedDateTimeISO("UTC");
    } else if (type.primitiveType == PrimitiveType.DATE) {
      return Temporal.PlainDate.from(value);
    } else if (type.primitiveType == PrimitiveType.TIME) {
      return Temporal.PlainTime.from(value);
    } else if (type.primitiveType == PrimitiveType.DURATION) {
      return timedeltaFromISOFormat(value);
    } else if (
      type.primitiveType == PrimitiveType.FLOAT16 ||
      type.primitiveType == PrimitiveType.FLOAT32 ||
      type.primitiveType == PrimitiveType.FLOAT64
    ) {
      return Number(value);
    } else if (
      type.primitiveType == PrimitiveType.SINT8 ||
      type.primitiveType == PrimitiveType.SINT16 ||
      type.primitiveType == PrimitiveType.SINT32 ||
      type.primitiveType == PrimitiveType.SINT64 ||
      type.primitiveType == PrimitiveType.SINT128 ||
      type.primitiveType == PrimitiveType.UINT8 ||
      type.primitiveType == PrimitiveType.UINT16 ||
      type.primitiveType == PrimitiveType.UINT32 ||
      type.primitiveType == PrimitiveType.UINT64 ||
      type.primitiveType == PrimitiveType.UINT128
    ) {
      return Number(value);
    } else if (type.primitiveType == PrimitiveType.JSON) {
      return value;
    } else {
      assertNever(type.primitiveType);
    }
  } else if (type.scalarType == ScalarType.ENUM) {
    return value;
  } else if (type.scalarType == ScalarType.NODE_REFERENCE) {
    const _NodeReference = STRUCT_CLASS_BY_TYPE[StructType.NODE_REFERENCE] as typeof NodeReference;
    return _NodeReference.unpack(Encoding.JSON, value, _session);
  } else if (type.scalarType == ScalarType.NODE_VALUE) {
    const nodeType = Number(value["type"]) as NodeType;
    const nodeClass = NODE_CLASS_BY_TYPE[nodeType];
    return nodeClass.unpack(Encoding.JSON, value, _session);
  } else if (type.scalarType == ScalarType.STRUCT) {
    if (type.structType === null) {
      throw new Error(`missing struct type for ${type.repr()}`);
    }
    const structClass = STRUCT_CLASS_BY_TYPE[type.structType];
    return structClass.unpack(Encoding.JSON, value, _session);
  } else {
    assertNever(type.scalarType);
  }
}
